Add support for lambda functions

andrewcleveland opened the pull request:
Adds to `create_function` a check if `func_name` is `'<lambda>'`, and if so builds a lambda expression instead of a regular function definition. This is passed in by `@wraps(f)` when `f` is a lambda function.
Without this, `create_function` builds a function definition `'def <lambda>(...):'` which fails to compile.
Also adds a test to verify the above is working.
Resolves #81.

`(lambda: None).__name__` is used in place of `'<lambda>'` in case a different Python implementation uses a different name.
Coverage.py complains about the `lambda: None` not being run. I don't know if that's an issue or not.

smarie commented on src/makefun/main.py:
```diff
@@ -197,6 +197,9 @@ def create_function(func_signature,             # type: Union[str, Signature]
         module_name = getattr_partial_aware(func_impl, '__module__', None)
 
+    # lambda check
+    is_lambda = func_name == (lambda: None).__name__
```
Would we instead rely on an independent util `is_lambda_function(f)` ? It would be easier to maintain in the long run.

Also `__code__.co_name` seems more reliable than `f.__name__` as it cannot be modified. See one of the comments in  https://stackoverflow.com/a/23852434/7262247

smarie commented on tests/test_advanced.py:
```diff
@@ -108,6 +108,19 @@ def goo(*args, **kwargs):
 
+def tests_wraps_lambda():
+    """ Tests that `@wraps` can duplicate the signature of a lambda """
+    foo = lambda a: a
```
Could you add a variant where `foo.__name__` is modified before calling `wraps` ? See above comment. Using `__code__.co_name` would fix the associated issue apparently.

smarie commented:
Thanks @andrewcleveland , this is definitely helpful ! I wrote a few comments, mostly details - the principle is good.

andrewcleveland commented:
Hi @smarie,
Per your comment about `__code__.co_name`, I've taken a different approach and added a new `co_name` parameter. A lambda is created only in the case that `co_name` is `'<lambda>'` or `func_impl.__code__.co_name` is not valid in a `def` statement.

smarie commented on src/makefun/_main_legacy_py.py:
```diff
@@ -88,3 +88,14 @@ def get_legacy_py_generator_body_template():
+
+def is_identifier(string):
+    """
+    Replacement for Python 3's `str.isidentifier`.
```
```suggestion
    Replacement for `str.isidentifier` when it is not available (e.g. on Python 2).
```

smarie commented on src/makefun/main.py:
```diff
@@ -225,11 +248,21 @@ def create_function(func_signature,             # type: Union[str, Signature]
                              "string containing the name" % func_impl)
-        func_signature_str = get_signature_string(func_name, func_signature, evaldict)
 
+        create_lambda = not _is_valid_func_def_name(co_name)
```
this seems to be a duplicate of the line above ?

smarie commented on src/makefun/main.py:
```diff
@@ -106,6 +107,7 @@ def create_function(func_signature,             # type: Union[str, Signature]
                     add_impl=True,              # type: bool
                     doc=None,                   # type: str
                     qualname=None,              # type: str
+                    co_name=None,               # type: str
```
Could you please update the body of the docstring to explain precisely the new behaviour ?
It is indeed hard to understand under which conditions lambdas are created vs. standard functions, as well as why users would like to change the new parameter.

smarie commented:
Thanks a lot @andrewcleveland ! I have two minor comments, mostly to improve readability/ future maintenance.

smarie commented:
Hi @andrewcleveland , would you mind updating your PR according to the two comments above, so that we can merge ? Users of `pytest-cases` will be happy about your mod :) (see https://github.com/smarie/python-pytest-cases/issues/278)

What is missing is just code comments in two sections.
Thanks in advance !

andrewcleveland commented:
Hi @smarie, sorry for the delay. I did some work on this about a month ago and then kind of put it on the back burner.
I think I've made the comment changes you requested but I may have missed something.
I also added the `co_name` parameter to some of the other functions for consistency.

smarie commented on src/makefun/main.py:
```diff
@@ -177,10 +194,25 @@ def create_function(func_signature,             # type: Union[str, Signature]
+    else:
+        if not (_is_valid_func_def_name(co_name)
+                or _is_lambda_func_name(co_name)):
+            raise ValueError("Invalid co_name %s for created function. "
```
```suggestion
            raise ValueError("Invalid co_name %r for created function. "
```

smarie replied:
Minor: quotes will help readability

smarie commented on src/makefun/main.py:
```diff
@@ -177,10 +194,25 @@ def create_function(func_signature,             # type: Union[str, Signature]
         user_provided_name = False
 
+    # co_name default
+    user_provided_co_name = True
+    if co_name is None:
+        if func_name is None:
+            co_name = '<lambda>'
+        else:
+            co_name = func_name
+        user_provided_co_name = False
```
Probably more readable:

```suggestion
    user_provided_co_name = co_name is not None
    if not user_provided_co_name :
        if func_name is None:
            co_name = '<lambda>'
        else:
            co_name = func_name
```

smarie commented on src/makefun/main.py:
```diff
@@ -11,6 +11,7 @@
 from inspect import getsource
+from keyword import iskeyword
 from textwrap import dedent
 from types import FunctionType
```
See comment below


```suggestion
from types import FunctionType


if sys.version_info >= (3, 0):
    is_identifier = str.isidentifier
else:
    from makefun._main_legacy_py import is_identifier
```

smarie commented on src/makefun/main.py:
```diff
@@ -297,6 +341,28 @@ def _is_generator_func(func_impl):
+    return func_name == (lambda: None).__code__.co_name
+
+
+def _is_valid_func_def_name(func_name):
+    """
+    Return True if func_name is valid in a function definition.
+    :param func_name:
+    :return:
+    """
+    if sys.version_info >= (3, 0):
+        is_identifier = str.isidentifier
+    else:
+        from makefun._main_legacy_py import is_identifier
+    return is_identifier(func_name) and not iskeyword(func_name)
```
Let's add the `is_identifier` symbol at the top of the module so that it is checked only once

```suggestion
def _is_valid_func_def_name(func_name):
    """
    Return True if func_name is valid in a function definition.
    :param func_name:
    :return:
    """
    return is_identifier(func_name) and not iskeyword(func_name)
```

smarie commented on src/makefun/main.py:
```diff
@@ -255,6 +294,8 @@ def create_function(func_signature,             # type: Union[str, Signature]
     elif isasyncgenfunction(func_impl):
         body = "async def %s\n    async for y in _func_impl_(%s):\n        yield y\n" % (func_signature_str, params_str)
+    elif create_lambda:
+        body = "lambda_ = lambda%s: _func_impl_(%s)\n" % (func_signature_str, params_str)
```
This could probably be the place where to add the extra space ? (see previous comment)

smarie commented on src/makefun/main.py:
```diff
@@ -208,28 +240,35 @@ def create_function(func_signature,             # type: Union[str, Signature]
     else:
         raise TypeError("Invalid type for `func_signature`: %s" % type(func_signature))
 
+    # add space to lambda func_signature_str, if not empty
+    if create_lambda and func_signature_str:
+        func_signature_str = " " + func_signature_str
```
Any reason why this is not done line 298 ? see below comment

smarie commented:
Thanks @andrewcleveland  ! I had a few other comments, hopefully easy to cope with

andrewcleveland commented:
Hi @smarie, I made the changes you suggested.
I changed the adding a space to `func_signature_str` to a check that uses a different `body` string for a lambda with/without arguments. It seems a bit easier to read.

Moving the conditional `is_identifier` import to the start of main.py seems to create a circular dependency..
The following error happens:
```
.nox/tests-2-7/lib/python2.7/site-packages/makefun/main.py:22: in <module>
     from makefun._main_legacy_py import is_identifier
 .nox/tests-2-7/lib/python2.7/site-packages/makefun/_main_legacy_py.py:8: in <module>
     from makefun.main import wraps
 E   ImportError: cannot import name wraps
```
I moved the import to the bottom of main.py, but I'm not sure if that's the best way to resolve it.

andrewcleveland commented:
@smarie, option 1 is done.

smarie commented:
All set, thanks a lot @andrewcleveland for your patience !
I'll tag and release asap
